illumos-utils/src/zfs.rs

```diff
@@ -375,6 +382,52 @@ impl fmt::Display for PropertySource {
     }
 }
 
+#[derive(Copy, Clone, Debug)]
+pub enum WhichDatasets {
+    SelfOnly,
+    SelfAndChildren,
+}
+
+// A helper structure for gathering all possible key/value pairs to pass to "zfs
+// set". This helps callers minimize the number of calls to "zfs set" they need
+// to make.
+struct PropertySetter {
+    props: BTreeMap<&'static str, String>,
+}
+
+impl PropertySetter {
+    fn new() -> Self {
+        PropertySetter { props: BTreeMap::new() }
+    }
+
+    fn add_size_details(&mut self, details: SizeDetails) -> &mut Self {
+        let SizeDetails { quota, reservation, compression } = details;
+        let quota = quota
+            .map(|q| q.to_bytes().to_string())
+            .unwrap_or_else(|| String::from("none"));
+        self.props.insert("quota", quota);
+
+        let reservation = reservation
+            .map(|r| r.to_bytes().to_string())
+            .unwrap_or_else(|| String::from("none"));
+        self.props.insert("reservation", reservation);
+
+        let compression = compression.to_string();
+        self.props.insert("compression", compression);
+
+        self
+    }
+
+    fn add_id(&mut self, id: DatasetUuid) -> &mut Self {
+        self.props.insert("oxide:uuid", id.to_string());
+        self
+    }
+
+    fn as_vec(&self) -> Vec<(&str, &str)> {
```

Comment:
Sure, I can pivot to that format. I do still want to keep this as a slightly more flexible `PropertySetter` rather than a function (who knows what properties we're going to want to add tomorrow).

I admittedly was using a BTreeMap to avoid setting a value twice, but I suppose we can just claim that's a user-error? `zfs set` explicitly returns an error in this case.

Either way: updating to avoid the extra data structure.